rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -29,264 +26,363 @@ trait HotStore[F[_], C, P, A, K] {
-    for {
-      state <- cacheRef.get
-      res <- state.joins.get(channel) match {
-              case None =>
-                for {
-                  historyJoins <- historyReader.getJoins(channel)
-                  _ <- cacheRef.update { cache =>
-                        ignore(cache.joins.putIfAbsent(channel, historyJoins))
-                        cache
-                      }
-                } yield (historyJoins)
-              case Some(joins) => joins.pure[F]
-            }
-    } yield res
+    } yield result
 
   def putJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
-      joins <- getJoins(channel)
-      _ <- cacheRef
-            .update { cache =>
-              ignore(cache.joins.put(channel, join +: joins))
-              cache
+      fromHistoryStore <- getJoinsFromHistoryStore(channel)
+      _ <- hotStoreState.update { state =>
+            val curJoins = state.joins.getOrElse(channel, fromHistoryStore)
+            if (curJoins.contains(join))
+              state
+            else {
+              val newVal = join +: curJoins
+              state.copy(joins = state.joins.updated(channel, newVal))
             }
-            .whenA(!joins.contains(join))
+          }
     } yield ()
 
-  def installJoin(channel: C, join: Seq[C]): F[Unit] = cacheRef.update { cache =>
-    ignore {
-      val installed = cache.installedJoins.getOrElse(channel, Seq.empty)
-
-      if (!installed.contains(join))
-        cache.installedJoins
-          .put(channel, join +: installed)
-    }
-    cache
+  def installJoin(channel: C, join: Seq[C]): F[Unit] = hotStoreState.update { state =>
+    val curInstalled = state.installedJoins.getOrElse(channel, Seq.empty)
+    if (!curInstalled.contains(join)) {
+      val newVal = join +: curInstalled
+      state.copy(installedJoins = state.installedJoins.updated(channel, newVal))
+    } else
+      state
   }
 
   def removeJoin(channel: C, join: Seq[C]): F[Unit] =
     for {
-      joins         <- internalGetJoins(channel)
-      continuations <- getContinuations(join)
-      s             <- cacheRef.get
-      index         = s.joins(channel).indexOf(join)
-      _ <- if (continuations.isEmpty && index != -1)
-            checkIndex(s.joins(channel), index) >>
-              cacheRef.update { cache =>
-                val updated =
-                  removeIndex(
-                    cache.joins(
-                      channel
-                    ),
-                    index
-                  )
-                ignore(
-                  cache.joins.put(
-                    channel,
-                    updated
-                  )
-                )
-                cache
-              } else ().pure[F]
+      joinsInHistoryStore <- getJoinsFromHistoryStore(channel)
+      contsInHistoryStore <- getContFromHistoryStore(join)
+      err <- hotStoreState.modify[Boolean](state => {
+
+              val curJoins = state.joins.getOrElse(channel, joinsInHistoryStore)
```

Comment:
@zsluedem I'm thinking - is here with joins situation the same as with continuations? `curJoin` here does not account installed.

Reply:
`removeJoin` got very specific target on removing

```
def removeJoin(channel: C, join: Seq[C])`
```

Not like continuations want `index`